test/lit/string.as_wtf16.wast

```diff
@@ -0,0 +1,119 @@
+;; NOTE: Assertions have been generated by update_lit_checks.py and should not be edited.
+
+;; Check that string.as_wtf16 is accepted by the parser but is not translated
+;; into any IR.
+
+;; RUN: wasm-opt %s -all -S -o - | filecheck %s
+;; RUN: wasm-opt %s -all --roundtrip -S -o - | filecheck %s --check-prefix=RTRIP
+
+(module
+  ;; CHECK:      (func $empty (type $1)
+  ;; CHECK-NEXT:  (nop)
+  ;; CHECK-NEXT: )
+  ;; RTRIP:      (func $empty (type $1)
+  ;; RTRIP-NEXT:  (nop)
+  ;; RTRIP-NEXT: )
+  (func $empty
+    (string.as_wtf16)
+  )
+
+  ;; CHECK:      (func $codeunit (type $0) (result i32)
+  ;; CHECK-NEXT:  (stringview_wtf16.get_codeunit
+  ;; CHECK-NEXT:   (string.const "abc")
+  ;; CHECK-NEXT:   (i32.const 0)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  ;; RTRIP:      (func $codeunit (type $0) (result i32)
+  ;; RTRIP-NEXT:  (local $0 i32)
+  ;; RTRIP-NEXT:  (local $1 (ref string))
```

Comment:
We should make an effort to avoid increasing code size with each unoptimized roundtrip. I ran into the same thing with `br_if` and the solution there was to only use scratch locals when we actually need them. I'm not sure offhand how to do that here.

Reply:
We always need them here, since there's no way around emitting the conversion on the input.

Reply:
That worries me, then. Unbounded increases in size are potentially annoying in things like the fuzzer.

Another possible way to do this is to mark ahead of time the first child in some data structure ("children that need to be stashed in a local"). Then in visitExpression when you hit such a child you stash it. That would avoid needing to stash both children. However, this has downsides - I tried it in my `br_if` work and found some issues, but they were specific to StackIR so perhaps it can work here. However I'd like to unify all of this later in StackIR, so perhaps not.

Reply:
Trying to optimize this would also become tricky in the face of multivalue. Consider the case where both stashed operands are produced by the same multivalue block or call. Right now that just works, but if we tried to handle the scratch locals on the producer rather than the consumer side, it would get very tricky indeed.

Reply:
Makes sense.

I think we can limit the growth, though. How about checking if the child we need to stash is a local.tee, and that that local is not written in the child after it? In that case we can avoid adding a new local and reuse that existing tee, so after one roundtrip we "stabilize".

Reply:
So if we have this:

```wasm
<produce string>
<produce start>
local.tee $x
<produce end>
local.tee $y
stringview_wtf16.slice
```

We can lower it to this:

```wasm
<produce string>
<produce start>
local.set $x
<produce end>
local.set $y
string.as_wtf16
local.get $x
local.get $y
stringview_wtf16.slice
```

But then when we parse it back in there are no tees, so this optimization does not limit IR growth from round tripping. Without limiting growth, it doesn't seem worth the complexity to me.

Reply:
(I think the tees can remain tees if we add drops; or, the pattern could support sets and not just tees?)

Reply:
The whole point of the scratch locals is to get the start and end values off the stack. If we use tees, then we don't get them off the stack and there is no way for the `string.as_wtf16` to consume the deeper string value.

Adding drops also doesn't help, since it would make the emitted code look like this:

```wasm
<produce string>
<produce start>
local.tee $x
drop
<produce end>
local.tee $y
drop
string.as_wtf16
local.get $x
local.get $y
stringview_wtf16.slice
```

Round tripping this would still have unbounded growth.

Reply:
I was able to optimize the operands when they are local.gets, and I was also able to avoid having to parse a `nop` in the binary reader. With those two changes, we now quickly reach a steady state for code size.